---
title: "Omics Coursework - Part 2, R Markdown"
author: "Andrej Hric"
date: "`r Sys.Date()`"
output: html_document
runtime: shiny
---

#Load libraries 
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

#libraries for Questions i-vi
library(DESeq2)        # for differential expression analysis using RNA-seq data
library(ggplot2)       # for creating complex and customizable data visualizations
library(tidyr)         # for tidying data, making it easier to manipulate
library(dplyr)         # for data manipulation and transformation
library(biomaRt)       #for querying Ensembl database

#libraries for shiny app Question vii
library(shiny)         # for building interactive web applications in R
library(shinythemes)   # for adding themes to shiny applications

```

#Question i) Differential expression analysis using DESeq2

```{r differential-expression-analysis}

#check and set correct working directory(adjust to ../input_data/p2  )
print(getwd())
setwd("/Users/andrejhric/Desktop/omics_coursework/input_data/p2")

#loading count data (containing gene expression counts for each sample) 
count_data <- read.table("GSE64018_countlevel_12asd_12ctl_edited.txt", header = TRUE, row.names = 1)
print("count data loaded correctly")

#loading metadata txt file
metadata <- read.table("metadata.txt", header = TRUE, row.names = 1)
print("metadata loaded correctly")

#create DESeqDataSet object
dds <- DESeqDataSetFromMatrix(countData = count_data, colData = metadata, design = ~ diagnosis)     

```

```
# Output 
  >Warning in DESeqDataSet(se, design = design, ignoreRank) :
    some variables in design formula are characters, converting to factors
```

```{r check-data-types}
#check variable types and convert character variables to factors
str(metadata)

metadata <- metadata %>%
  mutate(across(where(is.character), as.factor))

#check changes after converting
str(metadata)
```

```
# Original output (truncated)
  data.frame':	24 obs. of  22 variables:
   $ regionid                    : chr  "ba41-42-22" "ba41-42-22" "ba41-42-22" "ba41-42-22" ...
   $ diagnosis                   : chr  "ASD" "ASD" "ASD" "ASD" ...
   $ Sex                         : chr  "M" "M" "F" "F" ...
   $ brainbank                   : chr  "ATP" "ATP" "NICHD" "ATP" ...
 ```
 ```
 # Output after converting (truncated)
  'data.frame':	24 obs. of  22 variables:
   $ regionid                    : Factor w/ 1 level "ba41-42-22": 1 1 1 1 1 1 1 1 1 1 ...
   $ diagnosis                   : Factor w/ 2 levels "ASD","CTL": 1 1 1 1 1 1 1 1 1 1 ...
   $ Sex                         : Factor w/ 2 levels "F","M": 2 2 1 1 2 1 2 2 2 1 ...
   $ brainbank                   : Factor w/ 2 levels "ATP","NICHD": 1 1 2 1 1 1 1 1 1 1 ...
```

```{r check-data-types}
#rerun dds
dds <- DESeqDataSetFromMatrix(countData = count_data, colData = metadata, design = ~ diagnosis)     

#reveal diagnosis factor, by specifying reference level (set control group as reference)
dds$diagnosis <- relevel(dds$diagnosis, ref = "CTL")

#filter out low read counts (10 set as threshold commonly used in RNA-seq analysis to balance between removing noise but retaining potential genes of interest)
dds <- dds[rowSums(counts(dds)) > 10, ]

#run pipeline (this also normalizes data as DESeq internally estimates dispersions, size factors, and fits the model to perform the analysis)
dds <- DESeq(dds)
```
```  
#Output 
  estimating size factors
  estimating dispersions
  gene-wise dispersion estimates
  mean-dispersion relationship
  final dispersion estimates
  fitting model and testing
  -- replacing outliers and refitting for 558 genes
  -- DESeq argument 'minReplicatesForReplace' = 7 
  -- original counts are preserved in counts(dds)
  estimating dispersions
  fitting model and testing
```    
```{r show results}
#display results
res<- results(dds)
res
```  
```  
#Output

log2 fold change (MLE): diagnosis ASD vs CTL 
Wald test p-value: diagnosis ASD vs CTL 
DataFrame with 42839 rows and 6 columns
                 baseMean log2FoldChange     lfcSE       stat    pvalue      padj
                <numeric>      <numeric> <numeric>  <numeric> <numeric> <numeric>
ENSG00000000003  90.51022     0.02881461  0.224423  0.1283941  0.897837  0.957595
ENSG00000000005   2.39821     0.44433068  0.569226  0.7805879  0.435045  0.678601
ENSG00000000419 248.03262    -0.30067187  0.235277 -1.2779468  0.201268  0.458348
ENSG00000000457 218.69538    -0.18747117  0.147460 -1.2713367  0.203609  0.461427
ENSG00000000460  90.22595    -0.00792495  0.156331 -0.0506934  0.959570  0.984731
...                   ...            ...       ...        ...       ...       ...
ENSG00000272541  9.131107     -0.5903244  0.332623 -1.7747579 0.0759379  0.282732
ENSG00000272542 15.303447     -0.2188573  0.282689 -0.7741985 0.4388134  0.681977
ENSG00000272543 61.674685      0.2284879  0.238438  0.9582688 0.3379272  0.595390
ENSG00000272544  5.684451      0.5185789  0.375903  1.3795542 0.1677239  0.417281
ENSG00000272545  0.454525     -0.0114485  1.241902 -0.0092185 0.9926448        NA
```  
#Question ii) number of genes differentially expressed (up/down regulated) at adjusted p-value cut off of 0.05. = answer is 1236 in total

```{r}
#count differentially expressed genes with padj < 0.05
num_diff_expressed <- sum(res$padj < 0.05, na.rm=TRUE)  

#print result of counting 
cat("total number of differentially expressed genes (padj < 0.05):", num_diff_expressed, "\n")
```

```{r}
#count up and down regulated genes (padj < 0.05 and log2FoldChange > 0)
num_upregulated <- sum(res$padj < 0.05 & res$log2FoldChange > 0, na.rm=TRUE)

num_downregulated <- sum(res$padj < 0.05 & res$log2FoldChange < 0, na.rm=TRUE)

#print results
cat("number of upregulated genes (padj < 0.05):", num_upregulated, "\n")
cat("number of downregulated genes (padj < 0.05):", num_downregulated, "\n")
cat("total number of differentially expressed genes (padj < 0.05):", num_diff_expressed, "\n")
```

```
  Number of upregulated genes (padj < 0.05): 1169 
  Number of downregulated genes (padj < 0.05): 67 
  Total number of differentially expressed genes (padj < 0.05): 1236 
```

#Question iii) Volcano Plot

This volcano plot below displays differentially expressed genes. In the first volcano plot, genes that are statistically significant, with adjusted p-values less than 0.05, are shown in blue. The top 10 genes with the highest absolute fold change are labeled with their corresponding Ensembl IDs, and these genes are slightly larger for better visibility.

The second volcano plot highlights the upregulated and downregulated genes. Upregulated genes are depicted in orange, while downregulated genes are shown in purple. Additionally, this plot indicates that all of the top 10 genes were upregulated.

```{r volcano-plot, fig.width=10, fig.height=8}
#convert results to a data frame for ggplot2
res_df <- as.data.frame(res)

#remove rows with NA in the adjusted p-value column to avoid errors
res_df <- res_df %>% drop_na(padj)

#add column to indicate significant genes (adjusted p-value < 0.05)
res_df <- mutate(res_df, sig = ifelse(padj < 0.05, "significant", "not significant"))

#add column to indicate the top 10 signif genes
top_n <- 10
res_df <- res_df %>%
  arrange(desc(abs(log2FoldChange))) %>%
  mutate(top10 = ifelse(row_number() <= top_n, TRUE, FALSE))

#create volano plot
volcano_plot <- ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), color = sig)) +
  geom_point(size = 0.75) +
  scale_color_manual(values = c("black", "blue"), labels = c("not significant", "fdr<0.05")) +
  labs(x = "log2 fold change", y = "-log10 adjusted p-value", color = "significance") +
  theme_minimal() +
  theme(
    legend.position = "right",
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 12),  # increase the size of legend text
    legend.key.size = unit(1.5, "lines"),   # increase the size of legend points
    panel.grid.major = element_line(color = "grey80"),
    panel.grid.minor = element_line(color = "grey90"),
    panel.border = element_rect(color = "black", fill = NA, size = 1)
  ) +
  geom_text_repel(
    data = subset(res_df, top10 & sig == "significant"),
    aes(label = rownames(subset(res_df, top10 & sig == "significant"))),
    color = "black",
    size = 3
  )

#print output
print(volcano_plot)
```

```{r volcano-plot, fig.width=10, fig.height=8}
#all steps as before
res_df <- as.data.frame(res)
res_df <- res_df %>% drop_na(padj)

#add categorization of genes
res_df <- res_df %>%
  mutate(
    category = case_when(
      padj < 0.05 & log2FoldChange > 0 ~ "upregulated",
      padj < 0.05 & log2FoldChange < 0 ~ "downregulated",
      TRUE ~ "not significant"
    )
  )

top_n <- 10
res_df <- res_df %>%
  arrange(desc(abs(log2FoldChange))) %>%
  mutate(top10 = ifelse(row_number() <= top_n, TRUE, FALSE))

volcano_plot <- ggplot(res_df, aes(x = log2FoldChange, y = -log10(padj), color = category)) +
  geom_point(size = 0.75) +
  scale_color_manual(values = c("gray", "purple", "orange"), labels = c("not significant", "downregulated", "upregulated")) +
  labs(x = "log2 fold change", y = "-log10 adjusted p-value", color = "category") +
  theme_minimal() +
  theme(
    legend.position = "right",
    legend.title = element_text(size = 12),
    legend.text = element_text(size = 12),  # increase the size of legend text
    legend.key.size = unit(1.5, "lines"),   # increase the size of legend points
    panel.grid.major = element_line(color = "grey80"),
    panel.grid.minor = element_line(color = "grey90"),
    panel.border = element_rect(color = "black", fill = NA, size = 1)
  ) +
  geom_text_repel(
    data = subset(res_df, top10 & category != "not significant"),
    aes(label = rownames(subset(res_df, top10 & category != "not significant"))),
    color = "black",
    size = 3
  )

print(volcano_plot)

```

## Volcano Plot 1

![Volcano Plot 1](https://github.com/Andrej-Hric/Omics_coursework/output_data/p2/Rplot_volcanoplot1.png)

## Volcano Plot 2

![Volcano Plot 2](https://github.com/Andrej-Hric/Omics_coursework/blob/main/output_data/p2/Rplot_volcanoplot2.png)

#Question iv) SRRM4 expresison analysis

A) The code bellow produces a boxplot showing the normalized counts for the SRRM4 gene for all samples. This plot is also then compared against the original figure 7b form the reference paper by Irimia et al.

B) In this case I used normalized counts for the gene expression comparison as raw counts from sequencng  can have high variability and this could mislead me by not accurately reflecting the real differences between the samples either due to differences in sequencing depth or other factors.

Using normalized counts adjust for this variability which makes it easier to compare across different samples and makes the data interpretation more clear.

```{r}
#set seed for reproducibility
set.seed(123)

#check and set correct working directory (adjust to your path)
print(getwd())
setwd("/Users/andrejhric/Desktop/omics_coursework/input_data/p2")

#load count data (containing gene expression counts for each sample)
count_data <- read.table("GSE64018_countlevel_12asd_12ctl_edited.txt", header = TRUE, row.names = 1)
print("count data loaded successfully")

#load metadata txt file
metadata <- read.table("metadata.txt", header = TRUE, row.names = 1)
print("metadata loaded successfully")

#convert character metadata to factors
metadata <- metadata %>% mutate(across(where(is.character), as.factor))

#create DESeqDataSet object
dds <- DESeqDataSetFromMatrix(countData = count_data, colData = metadata, design = ~ diagnosis)
#set reference level
dds$diagnosis <- relevel(dds$diagnosis, ref = "CTL")

#filter out low count reads (less than 10 total reads)
dds <- dds[rowSums(counts(dds)) > 10, ]
print("low count reads filtered")

#add size factors and run DESeq
dds <- DESeq(dds)
print("DESeq analysis completed")

#create results table
res <- results(dds)
print("results extracted")

#convert results to a new dataframe and remove rows with "padj" NA values
df.plot <- as.data.frame(res) %>% drop_na(padj)

#order dataframe by absolute log2FoldChange values in descending order
df.plot <- df.plot[order(abs(df.plot$log2FoldChange), decreasing = TRUE), ]

#add column for significance label
df.plot <- mutate(df.plot, sig = ifelse(df.plot$padj < 0.05, "FDR<0.05", "Not sig"))

#add column for 10 most differentially expressed genes
top.n <- 10
df.plot <- cbind(df.plot, top10 = "no")
top10.index <- rownames(df.plot[df.plot$sig == "FDR<0.05", ][1:top.n, ])
df.plot[top10.index, "top10"] <- "yes"

#biomart database and dataset
ensembl <- useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")

#biomart query
g <- getBM(attributes = c("ensembl_gene_id", "external_gene_name"), filters = "ensembl_gene_id", mart = ensembl, values = rownames(df.plot))

#ensure the 'ensembl_id' column is correctly added to df.plot
df.plot$ensembl_id <- rownames(df.plot)

#merge the two dataframes ensuring the column names match
df.plot.gene <- merge(df.plot, g, by.x = "ensembl_id", by.y = "ensembl_gene_id")
colnames(df.plot.gene)[10] <- "gene_name"
print("gene names added to the dataframe")

#ensure SRRM4 is present
if("SRRM4" %in% df.plot.gene$gene_name) {
  print("SRRM4 gene is present.")
} else {
  print("SRRM4 gene is NOT present.")
}
```
``` 
#Output of tests !
[1] "count data loaded successfully"
[1] "metadata loaded successfully"
[1] "low count reads filtered"
estimating size factors
estimating dispersions
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
-- replacing outliers and refitting for 558 genes
-- DESeq argument 'minReplicatesForReplace' = 7 
-- original counts are preserved in counts(dds)
estimating dispersions
fitting model and testing
[1] "DESeq analysis completed"
[1] "results extracted"
[1] "gene names added to the dataframe"
[1] "SRRM4 gene is present."
```

```{r SRRM4, fig.height = 6, fig.width = 3, fig.align = "center"}
#check SRRM4 expression
srrm4_id <- df.plot.gene[df.plot.gene$gene_name == "SRRM4", ]$ensembl_id
if(length(srrm4_id) > 0) {
  
#extract normalized counts for SRRM4
srrm4_data <- plotCounts(dds, gene = srrm4_id, intgroup = "diagnosis", returnData = TRUE)
print(head(srrm4_data)) #check the data being plotted

#set seed for consistant boxplot so that if I run it is always the same !
set.seed(123)
q <- ggplot(srrm4_data, aes(diagnosis, count)) +
  geom_boxplot() +
  geom_jitter(aes(col = diagnosis), alpha = 0.4, width = 0.1) +
  labs(x = "Diagnosis", y = "Normalized count") +
  guides(col = "none") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))
print(q)
} else {
  print("SRRM4 gene ID not found.")
}

#save the plo
ggsave(filename = "/Users/andrejhric/Desktop/omics_coursework/output_data/p2/SRRM4_boxplot.png", plot = q, width = 5, height = 5)
```

#Question v) Result Comparison to original paper for SRRM4 gene

To look at the difference in SRRM4 expression between control and ASD groups, I used ANOVA (p = 0.178), Kruskal-Wallis (p = 0.2987), and two Sample t-test (p = 0.18), all showing no significant difference (p > 0.05). These p-values indicate that there is no detectable or significant differential expression between the two groups as it's not statistically significant. However, the difference between my and the original paper's results might be because they used more complex normalization methods and included additional covariates like sex, age, brain bank origin, RIN, and PCA components. They included these covariates in their model, whereas I only focused on gene expression based on diagnosis, just as suggested by the original figure 7b, which might have been my oversight.

Also, they had a larger sample size and used more complex statistical techniques

```{r}
#ANOVA test
anova_result <- aov(count ~ diagnosis, data = srrm4_data)
summary(anova_result)

#Kruskal-Wallis test
kruskal_test <- kruskal.test(count ~ diagnosis, data = srrm4_data)
kruskal_test

#t-test
t_test <- t.test(count ~ diagnosis, data = srrm4_data)
t_test

```

```
Df  Sum Sq Mean Sq F value Pr(>F)
diagnosis    1  603128  603128   1.937  0.178
Residuals   22 6849839  311356               

	Kruskal-Wallis rank sum test

data:  count by diagnosis
Kruskal-Wallis chi-squared = 1.08, df = 1, p-value = 0.2987


	Welch Two Sample t-test

data:  count by diagnosis
t = 1.3918, df = 19.057, p-value = 0.18
alternative hypothesis: true difference in means between group CTL and group ASD is not equal to 0
95 percent confidence interval:
 -159.6437  793.7455
sample estimates:
mean in group CTL mean in group ASD 
         1802.112          1485.061 
```



#Question vi) Possible imrpovements for my future analysis

Given that the samples originated from humans and differ biologically, and considering the variations observed in sequencing statistics, I could improve my method by incorporating additional covariates such as sex, age, RNA integrity number (RIN), and brain bank origin to account for confounding variables as they did in the original paper and this could imrpove the accuracy of the model by creating a multivariate regression model to acocunt for all of these covariates at the same time.

Next I could imrpove the readabiltiy of the data by removing low-queality samples and introduces quality control for this.
 
Also I could use the same more advanced normalization techniques as in the original paper, such as adjusted FPKMs via regression analysis, better accounts for sequencing depth and sample quality variability, increasing comparability. Increase the sample size as I said in the last answer in question v).


